server/src/instant/grouped_queue.clj

```diff
@@ -5,7 +5,38 @@
    (java.util.concurrent.atomic AtomicInteger)
    (clojure.lang PersistentQueue)))
 
-(def empty-q PersistentQueue/EMPTY)
+;; ----------- 
+;; inflight-queue
+
+(def persisted-q-empty PersistentQueue/EMPTY)
+
+(def inflight-queue-empty
+  {:pending persisted-q-empty
+   :working []})
+
+(defn inflight-queue-put [inflight-queue item]
+  (update inflight-queue :pending conj item))
+
+(defn inflight-queue-empty? [{:keys [pending working] :as _inflight-queue}]
+  (and (empty? pending) (empty? working)))
+
+(defn inflight-queue-pop [{:keys [pending working]}]
+  {:pending (pop pending)
+   :working (if-let [item (first pending)]
+              (conj working item)
+              working)})
+
+(defn inflight-queue-workset [{:keys [working]}]
+  working)
+
+(defn inflight-queue-workset-clear [inflight-queue]
+  (assoc inflight-queue :working []))
+
+(defn inflight-queue-peek-pending [{:keys [pending] :as _inflight-queue}]
+  (first pending))
+
+;; ----------- 
+;; grouped-queue
```

Comment:
I realize I could use a deftype to create a `GroupedQueue` as well; I worry about having an atom inside a deftype, but heck, why not